src/com/sap/piper/PiperGoUtils.groovy

```diff
@@ -58,8 +60,9 @@ class PiperGoUtils implements Serializable {
 
             }
         }
-
-        utils.stashWithMessage('piper-bin', 'failed to stash piper binary', 'piper')
+        def piperVersion = steps.sh returnStdout: true, script: "./${piperExecutable} version"
```

Comment:
We should avoid new "script exit code 1" failures ...
You could print a speaking error message in case of a failure ...

Reply:
@OliverNocon : Would't that mean we can't use the sh step anymore since that step behaves like that?

As already outline above it is also fine from my point of view to use the approach outlined by @CCFenner. That means we get one log entry per piper-go-bin call instead of only one for the whole pipeline run, but that is acceptable from my point of view.

Reply:
sure we can, and we have to.

Just wrap in a try-catch and use `error` with a more speaking text, as e.g. done here https://github.com/SAP/jenkins-library/blob/067eb0ba6ab8cf6f52577f67e4b70f8762db3758/vars/containerExecuteStructureTests.groovy#L136

Reply:
I would not add another flag for emitting an error. We should rather remove the number of  flags passed from groovy than increase the number.

I see following alternatives:
* We could look into calling the version retrieval in case of an error in `piperExecuteBin`.
    We have to make sure though that we do not hide the original error in case the version retrieval call fails ...
* Add the version to the log with every call of the piper binary in `verbose` mode. Could be added to the generated code and prevent another execution of the binary
* ...

Reply:
> I would not add another flag for emitting an error. 

@OliverNocon I guess you mean "for emitting the version" (?).

In case there is no new flag for the version desired. Why not simply emitting the version of the downloaded piper bin right after downloading and setting the executable flag. 

In case we use it from the groovy lib that means we have the version in the log. That helps us troubleshooting. That approach is rather simple. Simple approaches are typically good.

Now we can start making things more complicated:

> We could look into calling the version retrieval in case of an error in

But what shall we do in case we have to troubleshoot something which does not raise an error, but still leads to unexpected results?

> Add the version to the log with every call of the piper binary in `verbose` mode.

Customer provides a log from a run which was not launched in verbose mode. Now we have to ask for a version with verbose mode. Customer is maybe not that experienced and has difficulties to enable verbose mode ... That all means we need longer than needed for troubleshooting an issue compared to a situation where we simply have that one more log record.

I would like to ask again to emit the version simply to the log right after materializing the piper go binary. That simplifies troubleshooting and costs almost nothing.